File: Backtracking/NM4/Go_NM4/NM4.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const (
	max = 8 + 1
)

var (
	br      = bufio.NewReader(os.Stdin)
	bw      = bufio.NewWriter(os.Stdout)
	n, m    int
	numList [max]int
)

func combinationWithRepetition(cnt, cur int) {
	if cnt == m {
		for i := 0; i < m; i++ {
			fmt.Fprintf(bw, "%d ", numList[i])
		}
		fmt.Fprintln(bw)
		return
	}

	for i := cur; i <= n; i++ {
		numList[cnt] = i
		combinationWithRepetition(cnt+1, i)
	}
}

func main() {
	defer bw.Flush()

	fmt.Fscanln(br, &n, &m)

	combinationWithRepetition(0, 1)
}

File: Backtracking/NM4/C_NM4/NM4.c
#include <stdio.h>

#define MAX 8 + 1

int n, m;
int num_list[MAX];

void combination_with_repetition(int cnt, int cur)
{
	if (cnt == m)
	{
		for (int i = 0; i < m; i++)
		{
			printf("%d ", num_list[i]);
		}
		printf("\n");
		return;
	}

	for (int i = cur; i <= n; i++)
	{
		num_list[cnt] = i;
		combination_with_repetition(cnt + 1, i);
	}
}

int main()
{
	scanf_s("%d %d", &n, &m);

	combination_with_repetition(0, 1);

	return 0;
}
